### ISSUE TITLE
Multi-class segmentation

### ISSUE THREAD

**Issue by yenlunliao** [JiayuanWang-JW/YOLOv8-multi-task#10]
Hello,
I am interested in using the multi-class segmentation in the multi-task structure.
For example, a structure like this:

```
- [[15, 18, 21], 1, Detect, [10]]  
- [[45], 1, Segment, [1, 32, 256]]
- [[33], 1, Segment, [10, 32, 256]]

```
I found that the current code is not supportive to multi-class segmentation, and I would like to try to modify it and make it work.
Is there any hints about which part of the code should I start with or focus on?

Thanks.

**Comment by JiayuanWang-JW** [comment]
Thanks for your interest in our work. 

There are two ways you could achieve your goal. 
1. Share one segmentation neck and use different segmentation heads to achieve it, but this is not a multi-class segmentation strictly. 
2. You need to change the segmentation labels file and corresponding the read and get labels code (such as collate_fn, __getitem__ and related functions). Of course, you will meet lots of challenges when you change the low-level layer code, you have a responsibility to debug them and make sure your change is correct. Then change the loss function. After that, you also need to change the results-related functions. Finally, you need to change the val-related code following the training code you changed. 

To be honest, this is a big project if you follow the second way to implement it.

Good luck.

**Comment by wmjlincy** [comment]
hello,
Do you implement the Multi-class segmentation of the second way?

**Comment by yenlunliao** [comment]
I did not implement a general version, but there are some tips if you would like to use it in some specific case. Overlapping segment cannot be distinguished in such way.

1. Change `yolo/data/utils.py:178`
`
for i in range(len(segments)):
        seg_class = cls[i][0]   

        # mask = ms[i] * (i + 1)
        # masks = masks + mask
        # masks = np.clip(masks, a_min=0, a_max=i + 1)
        
        mask = ms[i] * seg_class
        masks = np.maximum(masks, mask)
`
To make the mask filled by the class_id
To do this, we need to pass the class information to function `polygons2masks_overlap`
`def polygons2masks(imgsz, cls, segments, downsample_ratio=1):`

2. Change `yolo/data/augment.py:848`:
`polygons2masks(imgsz, cls, segments, downsample_ratio=1)`

3. Revise `yolo/v8/DecSeg/train.py:304`:
`preds.shape[1] == 2` for binary segmentation, and you should implement the logic of multitask.
`
if preds.shape[1] == 2:
  # binary segmentation for id 10
  gt_masks = masks.unsqueeze(1).clamp_(max=1)
  neg_mask = 1 - gt_masks
  binary_mask = torch.cat((neg_mask, gt_masks), dim=1)
elif preds.shape[1] == 3:
  # multi class segmentation for id 11, 12
  masks_11 = masks.clone().unsqueeze(1)
  masks_11[masks_11 == 12] = 0
  masks_11[masks_11 == 11] = 1

  masks_12 = masks.clone().unsqueeze(1)
  masks_12[masks_12 == 11] = 0
  masks_12[masks_12 == 12] = 1  

  gt_masks = masks.clone().unsqueeze(1).clamp_(max=1)
  neg_mask = 1 - gt_masks

  binary_mask = torch.cat((neg_mask, masks_11, masks_12), dim=1)
`

4. Change the predict code `yolo/v8/engine/predict_multi.py:338`: (if needed)
Extract the mask and implement the demonstration function.

However, according to my experiment, multi-class segmentation with multi-task learning is not robust, which makes the difficulty of each task imbalance and hard to train. 

Glad to hear some suggestions if you are working on this topic.

**Comment by hu874** [comment]
我也想知道这么实现
